File: backend/src/modules/sentimento/domain/universe_at.py
"""`universe_at(ts, filtro)`: the point-in-time universe, with `s3_inferred` barred BY TYPE."""

# `SPEC-001` §3.7, `CA-F3-4`, plan `07` item 7.10 (`T-07.8`/`CST-62`). `SPEC-001` §3.7 fixes the
# general vocabulary as a three-member type-sum:
#
#   universe_source ∈ { snapshot, s3_inferred, premium_index_witness }
#
# and then, one line below it, disqualifies one of the three for exactly one purpose:
#
#   "universe_source = s3_inferred e INADMISSIVEL no caminho de decisao. Ele deduz existencia
#    do simbolo da existencia do arquivo -- fato conhecivel ~30,3h depois e so para simbolos
#    cujos arquivos continuam publicados: survivorship e lookahead na mesma coluna."
#
# `PRD-001` §CA-F0-1 explains WHY this is not a hypothetical: the daily `exchangeInfo` snapshot
# series (`T-02.1`) only started being captured on `2026-08-25` (one manual capture; `Q1` -- the
# cron that would make this a real series -- is still open). Every `ts` requested before that
# date has NO snapshot witness at all, ONLY the S3-derived one. Without a TYPE-level exclusion,
# the tempting shortcut ("no snapshot? use the S3 witness as if it decided") turns every
# cross-sectional read of the past into a retrospective read by construction -- exactly the
# defect `PRD-001` line 613 names: "resultado anterior a primeira data de snapshot sai rotulado
# 'universo retrospectivo (s3_inferred) -- nao e o universo conhecivel em t'", never silently
# promoted to a decided answer.
#
# ── THE TWO TYPES, AND WHY THERE ARE TWO ────────────────────────────────────────────────────
#
# `UniverseSource` is `SPEC-001` §3.7's own three members, spelled once, for callers that need
# to TAG a witness (this module does, on `UniverseAtResult.s3_witness_source`).
# `DecisiveUniverseSource` is the ADMISSIBLE subset `decide_universe_membership` accepts --
# `Literal["snapshot", "premium_index_witness"]`, which has no `s3_inferred` member AT ALL. A
# caller cannot construct a `Mapping[DecisiveUniverseSource, frozenset[str]]` with an
# `"s3_inferred"` key and pass `mypy --strict` (`backend/scripts/lint.sh`): the key type itself
# has no such variant to assign, the same technique `SurvivorshipVerdict`
# (`dump_survivorship.py`) uses to make a third verdict unrepresentable. This is BY TYPE, not
# validation -- there is no `if source == "s3_inferred": raise` anywhere in this module, and
# `test_universe_at.py`'s structural falsifier proves none is needed.
#
# ── UNION OF THE TWO TESTEMUNHAS, DIVERGENCE MARKED (SPEC-001 §3.7, "D-18") ────────────────
#
# `universe_at` never merges the two witnesses silently: `UniverseAtResult.symbols` is the
# UNION (decided ∪ s3-witness), and `divergence` (`compare_symbol_sets`, reused from
# `instrument_universe_snapshot.py` -- the same function `D2.3`/`D2.4` already use, per the
# handoff's instruction not to invent a third representation) names exactly which symbols only
# one side attests. When no snapshot is available for `ts` at all (the `2025-08-01` case, per
# `CA-F0-1`), `decided_symbols` is `frozenset()` -- not because the decision confirmed an empty
# universe, but because there was nothing admissible to decide FROM -- and `label` carries
# `RETROSPECTIVE_LABEL` so a caller cannot mistake "empty because unconfirmed" for "empty
# because measured".
#
# ── DOMAIN, NOT infra/use_cases (`ADR-016`, `Natureza`) ─────────────────────────────────────
#
# No socket, no clock, no file. The two witnesses arrive as already-loaded data the caller
# built with `instrument_universe_snapshot.build_instrument_rows` (the snapshot side) and
# whatever S3-survivorship-derived listing `T-07.2`'s vocabulary already names (the s3 side) --
# this module reads no second representation of either.

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Final, Literal

from src.modules.sentimento.domain.instrument_universe_snapshot import (
    InstrumentRow,
    SymbolSetDivergence,
    compare_symbol_sets,
)

UniverseSource = Literal["snapshot", "s3_inferred", "premium_index_witness"]

# The decision-path admissible subset -- EXCLUDES `s3_inferred` structurally (see module
# docstring). `decide_universe_membership` below is the ONLY function whose signature carries
# this type; nothing else in this module needs to talk about "an admissible source" at all.
DecisiveUniverseSource = Literal["snapshot", "premium_index_witness"]

SNAPSHOT: Final[DecisiveUniverseSource] = "snapshot"
PREMIUM_INDEX_WITNESS: Final[DecisiveUniverseSource] = "premium_index_witness"

# Typed `UniverseSource`, deliberately NOT `DecisiveUniverseSource` -- assigning this constant
# where a `DecisiveUniverseSource` is expected is exactly the misuse `mypy --strict` refuses.
S3_INFERRED: Final[UniverseSource] = "s3_inferred"

# `PRD-001` line 613, literal: "universo retrospectivo (s3_inferred) -- nao e o universo
# conhecivel em t". The CODE, per this repository's exception-message and identifier rules
# (`CLAUDE.md`, "idioma de identificador"), names the reason in English; the SPEC's own words
# are what the comment above quotes.
RETROSPECTIVE_LABEL: Final[str] = "retrospective_before_first_snapshot"


@dataclass(frozen=True)
class UniverseFilter:
    """A read-time filter over the point-in-time universe.

    `SPEC-001` §6/Q5, literal: "universo e filtro na LEITURA; contractType, underlyingSubType,
    venue_symbol persistidos por linha" -- this dataclass filters on the two of those three
    fields `InstrumentRow` (`T-02.1`) already stores; it adds no new column. `None` on a field
    means "do not filter on this axis", not "match rows where the field is absent" -- the two
    are different questions and this dataclass keeps them apart the way `InstrumentRow`'s own
    `None`-vs-`()` distinction does.
    """

    market: str | None = None
    underlying_sub_type: tuple[str, ...] | None = None

    def matches(self, row: InstrumentRow) -> bool:
        """Return whether `row` survives every axis this filter constrains."""
        if self.market is not None and row.market != self.market:
            return False
        return not (
            self.underlying_sub_type is not None
            and row.underlying_sub_type != self.underlying_sub_type
        )


NO_FILTER: Final[UniverseFilter] = UniverseFilter()


def decide_universe_membership(
    witnesses: Mapping[DecisiveUniverseSource, frozenset[str]],
) -> frozenset[str]:
    """Return the union of every symbol an ADMISSIBLE witness names -- the decision path.

    `witnesses` is keyed by `DecisiveUniverseSource`, which has no `s3_inferred` member: no
    call site can add an `s3_inferred`-sourced set to this mapping and still pass
    `mypy --strict`. An empty `witnesses` (no admissible source available for the requested
    `ts`) returns `frozenset()` -- the caller (`universe_at`) is the one that distinguishes
    "decided, and empty" from "nothing admissible to decide from" via `RETROSPECTIVE_LABEL`,
    because that distinction depends on WHY the mapping is empty, which this function is never
    told.
    """
    decided: frozenset[str] = frozenset()
    for symbols in witnesses.values():
        decided = decided | symbols
    return decided


@dataclass(frozen=True)
class UniverseAtResult:
    """The point-in-time universe at `ts`: the union of both testemunhas, divergence marked.

    `symbols` is `decided_symbols | s3_witness_symbols` -- never one side alone. `label` is
    `None` when a snapshot witness was available (the decision is confirmed) and
    `RETROSPECTIVE_LABEL` when it was not (`decided_symbols` is `frozenset()` for lack of an
    admissible witness, and `symbols` is carried entirely by the `s3_inferred` side).
    """

    ts: str
    symbols: frozenset[str]
    decided_symbols: frozenset[str]
    s3_witness_symbols: frozenset[str]
    divergence: SymbolSetDivergence
    label: str | None


def universe_at(
    ts: str,
    filtro: UniverseFilter | None = None,
    *,
    snapshot_rows: Sequence[InstrumentRow] | None = None,
    s3_witness_symbols: frozenset[str] = frozenset(),
) -> UniverseAtResult:
    """Return the universe vigente at `ts`, filtered by `filtro` -- `SPEC-001` §3.7, `D7.7`.

    `snapshot_rows=None` means "no `exchangeInfo` snapshot exists for `ts`" (the honest state
    for any `ts` before `T-02.1`'s series started, `CA-F0-1`) -- distinct from `snapshot_rows=()`
    ("a snapshot exists and its filtered projection is empty"), the same `None`-vs-empty
    distinction `InstrumentRow.underlying_sub_type` already relies on elsewhere in this
    component. Only the former yields `RETROSPECTIVE_LABEL`.

    `s3_witness_symbols` is the S3-survivorship-derived witness (`T-07.2`'s vocabulary) -- it is
    NEVER passed to `decide_universe_membership`; it only contributes to the union and to
    `divergence`, per `s3_inferred`'s exclusion from the decision path.
    """
    resolved_filtro = filtro if filtro is not None else NO_FILTER
    witnesses: dict[DecisiveUniverseSource, frozenset[str]] = {}
    if snapshot_rows is not None:
        witnesses[SNAPSHOT] = frozenset(
            row.symbol for row in snapshot_rows if resolved_filtro.matches(row)
        )
    decided_symbols = decide_universe_membership(witnesses)
    return UniverseAtResult(
        ts=ts,
        symbols=decided_symbols | s3_witness_symbols,
        decided_symbols=decided_symbols,
        s3_witness_symbols=s3_witness_symbols,
        divergence=compare_symbol_sets(decided_symbols, s3_witness_symbols),
        label=None if snapshot_rows is not None else RETROSPECTIVE_LABEL,
    )

File: frontend/src/app/universe-at.ts
/**
 * `T-07.14` — TypeScript port of `universe_at(ts, filter)`, plus a NEW extension this task
 * adds (the delisting badge) that has no Python counterpart.
 *
 * `S5` is not a screen (`SPEC-001` §6, `docs/product/STITCH_CONTEXT.md:136`): it is
 * `universe_at(ts, filtro)` embedded behind whatever symbol selector another screen builds
 * (`docs/context/plataforma-dados/gates/T-07.14-design.md`: no screen has a real selector
 * built yet, so there is no UI decision pending here — this module ships the data contract
 * only, same posture `threshold-spec-bundle.ts` (`T-08.5`) took for `S4`).
 *
 * NORMATIVE SOURCE — read, never edited (`backend/` is off limits per `CLAUDE.md`):
 * `backend/src/modules/sentimento/domain/universe_at.py`. It is pure domain logic (`ADR-016`,
 * "Natureza": no socket, no clock, no file) — reimplementing the algorithm here, citing the
 * source, is the sanctioned path (same family as `T-08.5`'s own precedent comment); there is
 * no CLI to shell out to (`T-07.8`'s own builder handoff: "não escreve o use_case/CLI que lê
 * snapshot real e chama `universe_at`").
 *
 * THE CENTRAL RULE THIS PORTS, verbatim from the Python docstring: `UniverseSource` has three
 * members (`SPEC-001` §3.7), but `DecisiveUniverseSource` — the type
 * `decideUniverseMembership` accepts — has only two, `s3_inferred` EXCLUDED BY TYPE. Python
 * enforces this with `mypy --strict` on a `Literal` union; TypeScript has no `mypy --strict`
 * gate wired into this repository's `frontend/eslint.config.mjs` (no type-aware, no
 * `tsc --noEmit` script exists today — `[MEDIDO 2026-09-02: grep -n '"typecheck"\|project:'
 * frontend/package.json frontend/eslint.config.mjs → 0 linhas]`), but the discriminated union
 * still plays the same role a caller relies on: no value typed `DecisiveUniverseSource` can be
 * the string `"s3_inferred"`, so `decideUniverseMembership`'s parameter type has no key to
 * assign one under. Verified once by hand, the same division of labour
 * `test_universe_at.py`'s own comment describes for `mypy` (no `tsconfig.json` exists in
 * `frontend/` today — `[MEDIDO 2026-09-02: ls frontend/tsconfig.json → No such file or
 * directory]` — hence the explicit `--lib`/`--target` below instead of a project file):
 *
 *   $ printf 'import type { DecisiveUniverseSource } from "./universe-at.ts";\n' \
 *       'const bad: DecisiveUniverseSource = "s3_inferred";\n' > /tmp/scratch_bad.ts
 *   $ node_modules/.bin/tsc --noEmit --strict --lib es2020 --target es2020 \
 *       --moduleResolution bundler /tmp/scratch_bad.ts
 *   src/app/universe-at.ts(...): error TS2322: Type '"s3_inferred"' is not assignable to
 *   type 'DecisiveUniverseSource'.
 *   `[MEDIDO 2026-09-02, rodado de dentro de frontend/]`
 *
 * `universe-at.test.ts` proves the RUNTIME half — the same falsifier discipline
 * `test_universe_at.py` applies via `ast`, done here via `Function.prototype.toString()`
 * (source text survives Node's type-stripping, unlike Python's `inspect.getsource`, which is
 * why this reaches for the JS-native tool instead of porting the `ast` module's approach).
 *
 * THE EXTENSION — the delisting badge, which `T-07.8` never built:
 * `InstrumentRow` (backend `T-02.1`) has NO `deliveryDate` field
 * (`backend/src/modules/sentimento/domain/instrument_universe_snapshot.py`, grepped, absent;
 * confirmed again by `docs/context/plataforma-dados/handoff/T-07.14.md`, "`InstrumentRow`
 * ... NÃO tem campo `deliveryDate`"), so the badge cannot come from the same rows
 * `universeAt` decides membership from — it has to read the RAW `exchangeInfo` capture
 * directly (`data/snapshots/2026-08-25_exchangeInfo.json`, catalogued in `data/MANIFEST.md`).
 * `DELISTING_SENTINEL_DELIVERY_DATE` is the value measured directly in that file — 743/877
 * entries in the raw file carry it, and every one of `ICXUSDT`/`STORJUSDT`/`SCRTUSDT` carries
 * `1787734800000` instead (`[MEDIDO 2026-09-02: python3 -c "... Counter(s['deliveryDate'] for
 * s in json.load(open('data/snapshots/2026-08-25_exchangeInfo.json'))['symbols'])"]`, see the
 * QA gate block for the literal command). Do not re-derive the sentinel by any other path.
 *
 * `stampUniverseRows` is the "`universe_source` carimbado em toda saída" requirement, literal
 * from the task title: every row this module hands to a caller carries the union of
 * `UniverseSource` values that attested the symbol (never a silent merge — the same
 * `divergence`-is-data posture `universe_at.py` already takes), plus the badge.
 */

// ── The three-member vocabulary, and the two-member DECISIVE subset (SPEC-001 §3.7) ────────

export type UniverseSource = "snapshot" | "s3_inferred" | "premium_index_witness";

/**
 * The admissible subset `decideUniverseMembership` accepts — no `s3_inferred` member AT ALL.
 * See the module docstring above for why this is a type-level fence, not a runtime `if`.
 */
export type DecisiveUniverseSource = "snapshot" | "premium_index_witness";

export const SNAPSHOT: DecisiveUniverseSource = "snapshot";
export const PREMIUM_INDEX_WITNESS: DecisiveUniverseSource = "premium_index_witness";
/** Typed `UniverseSource`, deliberately NOT `DecisiveUniverseSource` — mirrors `universe_at.py`. */
export const S3_INFERRED: UniverseSource = "s3_inferred";

/**
 * `PRD-001` line 613, literal (quoted by `universe_at.py` too): "universo retrospectivo
 * (s3_inferred) — não é o universo conhecível em t".
 */
export const RETROSPECTIVE_LABEL = "retrospective_before_first_snapshot";

// ── The minimal snapshot-witness row this module's algorithm actually reads ────────────────
//
// This is NOT a full port of `InstrumentRow` (`instrument_universe_snapshot.py` has three more
// fields — `fundingIntervalHours`, `interestRate` — that neither `UniverseFilter` nor
// `universeAt` ever reads). Porting fields nothing here consumes would be scope this task does
// not own; `buildSnapshotWitnessRows` below joins only the two sources `market` derivation
// needs (`exchangeInfo`, `fundingInfo`), the same join `build_instrument_rows` performs.

export const MARKET_USDS_M = "USDS_M";
export const MARKET_COIN_M = "COIN_M";

export interface SnapshotWitnessRow {
  readonly symbol: string;
  readonly market: string;
  /** `null` = no `exchangeInfo` row this capture (an absence of OBSERVATION, not of tag). */
  readonly underlyingSubType: readonly string[] | null;
}

// ── `UniverseFilter` — a read-time filter, `SPEC-001` §6/Q5 ────────────────────────────────

export interface UniverseFilter {
  /** `undefined` = "does not filter this axis", never "match rows with no market". */
  readonly market?: string;
  readonly underlyingSubType?: readonly string[];
}

export const NO_FILTER: UniverseFilter = {};

function sameSequence(a: readonly string[], b: readonly string[]): boolean {
  return a.length === b.length && a.every((value, index) => value === b[index]);
}

/** Return whether `row` survives every axis `filter` constrains. */
export function filterMatches(filter: UniverseFilter, row: SnapshotWitnessRow): boolean {
  if (filter.market !== undefined && row.market !== filter.market) {
    return false;
  }
  if (filter.underlyingSubType !== undefined) {
    if (row.underlyingSubType === null || !sameSequence(row.underlyingSubType, filter.underlyingSubType)) {
      return false;
    }
  }
  return true;
}

// ── `SymbolSetDivergence` / `compareSymbolSets` — "a divergência é dado, não erro" ─────────

export interface SymbolSetDivergence {
  readonly onlyInFirst: readonly string[];
  readonly onlyInSecond: readonly string[];
}

export function compareSymbolSets(first: ReadonlySet<string>, second: ReadonlySet<string>): SymbolSetDivergence {
  const onlyInFirst = [...first].filter((symbol) => !second.has(symbol)).sort();
  const onlyInSecond = [...second].filter((symbol) => !first.has(symbol)).sort();
  return { onlyInFirst, onlyInSecond };
}

// ── `decideUniverseMembership` — the union of ADMISSIBLE witnesses only ────────────────────

export function decideUniverseMembership(
  witnesses: Partial<Record<DecisiveUniverseSource, ReadonlySet<string>>>,
): ReadonlySet<string> {
  const decided = new Set<string>();
  for (const symbols of Object.values(witnesses)) {
    if (symbols === undefined) {
      continue;
    }
    for (const symbol of symbols) {
      decided.add(symbol);
    }
  }
  return decided;
}

// ── `UniverseAtResult` / `universeAt` ───────────────────────────────────────────────────────

export interface UniverseAtResult {
  readonly ts: string;
  /** ALWAYS the union `decidedSymbols ∪ s3WitnessSymbols` — never one side alone. */
  readonly symbols: ReadonlySet<string>;
  readonly decidedSymbols: ReadonlySet<string>;
  readonly s3WitnessSymbols: ReadonlySet<string>;
  readonly divergence: SymbolSetDivergence;
  /** `null` when a snapshot witness was available; `RETROSPECTIVE_LABEL` when it was not. */
  readonly label: string | null;
}

export interface UniverseAtOptions {
  /**
   * `null`/omitted = "no `exchangeInfo` snapshot exists for `ts`" — distinct from `[]` ("a
   * snapshot exists and its filtered projection is empty"). Only the former yields
   * `RETROSPECTIVE_LABEL`, mirroring `universe_at.py`'s own `None`-vs-`()` distinction.
   */
  readonly snapshotRows?: readonly SnapshotWitnessRow[] | null;
  /** The S3-survivorship-derived witness. NEVER passed to `decideUniverseMembership`. */
  readonly s3WitnessSymbols?: ReadonlySet<string>;
}

export function universeAt(
  ts: string,
  filter: UniverseFilter | null = null,
  options: UniverseAtOptions = {},
): UniverseAtResult {
  const resolvedFilter = filter ?? NO_FILTER;
  const snapshotRows = options.snapshotRows ?? null;
  const s3WitnessSymbols = options.s3WitnessSymbols ?? new Set<string>();

  const witnesses: Partial<Record<DecisiveUniverseSource, ReadonlySet<string>>> = {};
  if (snapshotRows !== null) {
    witnesses[SNAPSHOT] = new Set(
      snapshotRows.filter((row) => filterMatches(resolvedFilter, row)).map((row) => row.symbol),
    );
  }
  const decidedSymbols = decideUniverseMembership(witnesses);
  const symbols = new Set<string>([...decidedSymbols, ...s3WitnessSymbols]);
  return {
    ts,
    symbols,
    decidedSymbols,
    s3WitnessSymbols,
    divergence: compareSymbolSets(decidedSymbols, s3WitnessSymbols),
    label: snapshotRows !== null ? null : RETROSPECTIVE_LABEL,
  };
}

// ── `buildSnapshotWitnessRows` — the join `SnapshotWitnessRow` needs, real-data plumbing ───

export interface RawExchangeInfoEntry {
  readonly symbol: string;
  readonly underlyingSubType?: readonly string[];
  readonly deliveryDate: number;
}

export interface RawExchangeInfoPayload {
  readonly symbols: readonly RawExchangeInfoEntry[];
}

export interface RawFundingInfoEntry {
  readonly symbol: string;
}

/**
 * Join `exchangeInfo` + `fundingInfo` into `SnapshotWitnessRow`s — the same union
 * `build_instrument_rows` computes (`exchangeInfo` ∪ `fundingInfo`), narrowed to the fields
 * `filterMatches`/`universeAt` read. Present in `exchangeInfo` ⇒ `MARKET_USDS_M`; present in
 * `fundingInfo` alone ⇒ `MARKET_COIN_M` (the COIN-M "stowaways", `instrument_universe_
 * snapshot.py`'s own term).
 */
export function buildSnapshotWitnessRows(
  exchangeInfo: RawExchangeInfoPayload,
  fundingInfo: readonly RawFundingInfoEntry[],
): readonly SnapshotWitnessRow[] {
  const exchangeBySymbol = new Map(exchangeInfo.symbols.map((entry) => [entry.symbol, entry]));
  const universe = new Set<string>([...exchangeBySymbol.keys(), ...fundingInfo.map((entry) => entry.symbol)]);
  return [...universe].sort().map((symbol) => {
    const exchangeEntry = exchangeBySymbol.get(symbol);
    return {
      symbol,
      market: exchangeEntry !== undefined ? MARKET_USDS_M : MARKET_COIN_M,
      underlyingSubType: exchangeEntry !== undefined ? (exchangeEntry.underlyingSubType ?? []) : null,
    };
  });
}

// ── The delisting badge — NEW in this task, no Python counterpart ──────────────────────────

/**
 * The sentinel value in `data/snapshots/2026-08-25_exchangeInfo.json` — a `deliveryDate`
 * different from this one is a REAL scheduled delisting, never a value to re-derive by
 * inference. Measured directly on the raw file (all 877 entries, every `contractType`/
 * `status`), not assumed:
 * `[MEDIDO 2026-09-02: python3 -c "import json,collections; d=json.load(open(
 * 'data/snapshots/2026-08-25_exchangeInfo.json')); print(len(d['symbols']),
 * collections.Counter(s['deliveryDate'] for s in d['symbols'])[4133404800000])" →
 * 877 símbolos totais, 743 com o sentinela]`. Restricted to `status=TRADING` +
 * `contractType=PERPETUAL` (`docs/recorte-plataforma.md:90`'s own "568/570" framing —
 * `[MEDIDO 2026-09-02: mesmo comando com filtro adicional 'status'=='TRADING' and
 * 'contractType'=='PERPETUAL' → 570 símbolos, 567 com o sentinela, 3 sem: ICXUSDT, STORJUSDT,
 * SCRTUSDT]`), the task's own anchor case — this module does not apply that restriction (it
 * has no `contractType`/`status` field to filter on; `stampUniverseRows` reads whatever
 * `exchangeInfo` entries the caller passes), it only measures the value the sentinel carries.
 */
export const DELISTING_SENTINEL_DELIVERY_DATE = 4133404800000;

export function hasScheduledDelisting(deliveryDate: number): boolean {
  return deliveryDate !== DELISTING_SENTINEL_DELIVERY_DATE;
}

export function deliveryDateBySymbol(payload: RawExchangeInfoPayload): ReadonlyMap<string, number> {
  const map = new Map<string, number>();
  for (const entry of payload.symbols) {
    map.set(entry.symbol, entry.deliveryDate);
  }
  return map;
}

/**
 * One output row of the "S5 embutido" contract: `symbol` + `universeSource` STAMPED (never
 * absent — the task's own title, literal) + the delisting badge. `universeSource` is the FULL
 * set of witnesses that attested the symbol — `["snapshot", "s3_inferred"]` when both did, per
 * the handoff ("cada símbolo devolvido carrega de qual fonte ele veio (snapshot, s3_inferred
 * ou ambos, via divergence)").
 */
export interface UniverseAtRow {
  readonly symbol: string;
  readonly universeSource: readonly UniverseSource[];
  /** `null` when no raw `exchangeInfo` entry exists for this symbol — badge stays honestly `false`. */
  readonly deliveryDate: number | null;
  readonly delistingBadge: boolean;
}

/**
 * Combine a `UniverseAtResult` with a raw `exchangeInfo` capture into per-symbol output rows.
 * `exchangeInfo=null` is the retrospective case (`RETROSPECTIVE_LABEL`, `CA-F0-1`): before
 * `T-02.1`'s first capture there is no raw payload to read a `deliveryDate` from at all, so
 * every row's badge is `false` — a `s3_inferred`-only symbol has never been observed for
 * delisting, and this function does not guess.
 */
export function stampUniverseRows(
  result: UniverseAtResult,
  exchangeInfo: RawExchangeInfoPayload | null,
): readonly UniverseAtRow[] {
  const deliveryDates = exchangeInfo !== null ? deliveryDateBySymbol(exchangeInfo) : new Map<string, number>();
  return [...result.symbols].sort().map((symbol) => {
    const universeSource: UniverseSource[] = [];
    if (result.decidedSymbols.has(symbol)) {
      universeSource.push(SNAPSHOT);
    }
    if (result.s3WitnessSymbols.has(symbol)) {
      universeSource.push(S3_INFERRED);
    }
    const deliveryDate = deliveryDates.get(symbol) ?? null;
    return {
      symbol,
      universeSource,
      deliveryDate,
      delistingBadge: deliveryDate !== null && hasScheduledDelisting(deliveryDate),
    };
  });
}
